openmmtools/storage/storageinterface.py

```diff
@@ -0,0 +1,521 @@
+        error.
+
+        Returns
+        -------
+        data
+            Data stored on VARIABLE read from disk and processed through the STORAGESYSTEM back into a Python type
+            and possibly through the UNIT logic to recast into Quantity before being handed to the user.
+        """
+        if not self.bound_target:
+            self._bind_read()
+        if not self._variable:
+            path = self.path
+            # Try to get variable on disk
+            try:
+                self._variable = self._storage_driver.get_variable_handler(path)
+            except KeyError as e:
+                raise e
+            self._directory = False
+        return self._variable.read()
+
+    def add_metadata(self, name, data):
+        """
+        Attempt to add metadata to the variable/directory. Usually a string input to include additional information
+        about the variable.
+
+        Because this can act on DIRECTORY or VARIABLE types, all metadata is buffered until this instance of SIDV is
+        bound to an object on the file itself.
+
+        This does not check if the metadata already exists and simply overwrites it if present.
+
+
+        Parameters
+        ----------
+        name : string
+            Name of the metadata variable. Since this is data we are directly attaching to the variable or directory
+            itself, we do not make this another SIDV instance.
+        data: What data you wish to attach to the `name`d metadata pointer.
+
+        """
+        bound_target = self.bound_target
+        if bound_target is None:
+            self._metadata_buffer[name] = data
+        else:
+            self._storage_driver.add_metadata(name, data, path=self.path)
+
+    """
+    PROTECTED FLAGS
+    Used for functionality checks that will likely never be called by the user, but exposed to show
+    what names can NOT be used in the dynamic directory/variable namespace a user will generate.
+    """
+    @property
+    def variable(self):
+        """
+        Checks if the object can be used in the .write, .append, .fetch functions can be used. Once this is set, this
+        instance cannot be converted to a directory type.
+
+        Returns
+        -------
+        variable_pointer : None or storage_system specific unit of storage
+            Returns None if this instance is a directory, or if its functionality has not been determined yet (this
+                includes the variable not being assigned yet)
+            Returns the storage_system specific variable that this instance is bound to once assigned
+
+        """
+        return self._variable
+
+    @property
+    def directory(self):
+        """
+        Checks if the object can be used as a directory for accepting other SIDV objects. Once this is set, the
+        .write, .append, and .fetch functions are locked out.
+
+        Returns
+        -------
+        directory_pointer : None, or storage_system specific directory of storage
```

Comment:
I would agree with this except for the part where there is a 3rd state: `unbound`. When no IO has actually taken place yet, the SIDV does not know if it is a variable or a directory, it could be either. The resulting logic would require me to either return `None` or some informative string as a 3rd option and coding around that. By having both `.variable` and `.directory` return None or the target its bound to, it lets me manipulate this 3rd option. 

The `bound_target` object still acts like you want it and the `.variable` and `.directory` variables will probably not be used much in practice, but are there if people want them. I can still refactor it if you think its a good idea.